File: scala/cargo-tracker/model2/src/main/scala/models/delivery/Delivery.scala
package models.delivery

import cats.{Id, ~>}
import cats.data.{Kleisli, State}
import cats.free.Free.liftF

import java.time.LocalDateTime

type TrackingId = String
type UnLocode = String
type VoyageNo = String
type Date = LocalDateTime

enum Delivery:
  case NotReceived(trackingId: TrackingId)
  case InPort(trackingId: TrackingId, location: UnLocode)
  case OnBoardCarrier(trackingId: TrackingId, currentVoyageNo: VoyageNo, location: UnLocode)
  case Claimed(trackingId: TrackingId, location: UnLocode, claimedTime: Date)

enum HandlingEvent:
  case Received(trackingId: TrackingId, location: UnLocode, completionTime: Date)
  case Loaded(trackingId: TrackingId, voyageNo: VoyageNo, location: UnLocode, completionTime: Date)
  case Unloaded(trackingId: TrackingId, voyageNo: VoyageNo, location: UnLocode, completionTime: Date)
  case Claimed(trackingId: TrackingId, location: UnLocode, completionTime: Date)

enum CommandA[A]:
  case Receive(location: UnLocode, completionTime: Date) extends CommandA[List[HandlingEvent]]
  case Load(voyageNo: VoyageNo, completionTime: Date) extends CommandA[List[HandlingEvent]]
  case Unload(location: UnLocode, completionTime: Date) extends CommandA[List[HandlingEvent]]
  case Claim(completionTime: Date) extends CommandA[List[HandlingEvent]]

type DeliveryState[A] = State[Delivery, A]

object Delivery:
  import CommandA.*

  def createDelivery(trackingId: TrackingId) = Delivery.NotReceived(trackingId)

  def receive(location: UnLocode, completionTime: Date) =
    liftF(Receive(location, completionTime))

  def load(voyageNo: VoyageNo, completionTime: Date) =
    liftF(Load(voyageNo, completionTime))

  def unload(location: UnLocode, completionTime: Date) =
    liftF(Unload(location, completionTime))

  def claim(completionTime: Date) =
    liftF(Claim(completionTime))

  def action[A](cmd: CommandA[A]): Kleisli[Id, Delivery, (Delivery, A)] = cmd match
    case Receive(l, c) => Kleisli {
      case NotReceived(t) =>
        (InPort(t, l), List(HandlingEvent.Received(t, l, c)))
      case s => (s, List.empty[HandlingEvent])
    }
    case Load(v, c) => Kleisli {
      case InPort(t, l) if v.nonBlank  =>
        (OnBoardCarrier(t, v, l), List(HandlingEvent.Loaded(t, v, l, c)))
      case s => (s, List.empty[HandlingEvent])
    }
    case Unload(l, c) => Kleisli {
      case OnBoardCarrier(t, v, _) =>
        (InPort(t, l), List(HandlingEvent.Unloaded(t, v, l, c)))
      case s => (s, List.empty[HandlingEvent])
    }
    case Claim(c) => Kleisli {
      case InPort(t, l) =>
        (Claimed(t, l, c), List(HandlingEvent.Claimed(t, l, c)))
      case s => (s, List.empty[HandlingEvent])
    }
  end action

  def interpret: CommandA ~> DeliveryState =
    new (CommandA ~> DeliveryState) {
      def apply[A](cmd: CommandA[A]): DeliveryState[A] = State {
        action(cmd).run(_)
      }
    }

  extension (s: VoyageNo)
    def nonBlank: Boolean = !s.isBlank

end Delivery

File: rust/cargo-tracker/sample/delivery/src/delivery.rs

use chrono::prelude::*;
use serde::{Deserialize, Serialize};
use std::fmt;

pub type TrackingId = String;
pub type UnLocode = String;
pub type VoyageNo = String;
pub type Date = DateTime<Local>;

#[derive(Debug, Clone, Deserialize, Serialize)]
pub enum Event {
    Created { tracking_id: TrackingId },
    Received { tracking_id: TrackingId, location: UnLocode, completion_time: Date },
    Loaded { tracking_id: TrackingId, voyage_no: VoyageNo, location: UnLocode, completion_time: Date },
    Unloaded { tracking_id: TrackingId, voyage_no: VoyageNo, location: UnLocode, completion_time: Date },
    Claimed { tracking_id: TrackingId, completion_time: Date },
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub enum Delivery {
    Nothing,
    NotReceived { tracking_id: TrackingId },
    InPort { tracking_id: TrackingId, location: UnLocode },
    OnBoardCarrier { tracking_id: TrackingId, voyage_no: VoyageNo, location: UnLocode },
    Claimed { tracking_id: TrackingId, location: UnLocode, claimed_time: Date },
}

#[derive(Debug, Clone)]
pub enum Command {
    Create(TrackingId),
    Receive(UnLocode, Date),
    Load(VoyageNo, Date),
    Unload(UnLocode, Date),
    Claim(Date),
}

#[derive(Debug, Clone)]
pub enum CommandError {
    InvalidState,
}

impl fmt::Display for CommandError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::InvalidState => write!(f, "invalid state"),
        }
    }
}

impl std::error::Error for CommandError {}

pub type DeliveryResult<T> = std::result::Result<T, CommandError>;

impl Delivery {
    pub fn action(&self, cmd: &Command) -> DeliveryResult<(Self, Event)> {
        match self {
            Self::Nothing => match cmd {
                Command::Create(t) => {
                    Ok((
                        Self::NotReceived { tracking_id: t.clone() },
                        Event::Created { tracking_id: t.clone() }
                    ))
                }
                _ => invalid_state()
            }
            Self::NotReceived { tracking_id } => match cmd {
                Command::Receive(l, d) => {
                    Ok((
                        Self::InPort { tracking_id: tracking_id.clone(), location: l.clone() },
                        Event::Received { tracking_id: tracking_id.clone(), location: l.clone(), completion_time: d.clone() }
                    ))
                }
                _ => invalid_state()
            }
            Self::InPort { tracking_id, location } => match cmd {
                Command::Load(v, d) => {
                    Ok((
                        Self::OnBoardCarrier { tracking_id: tracking_id.clone(), voyage_no: v.clone(), location: location.clone() },
                        Event::Loaded { tracking_id: tracking_id.clone(), voyage_no: v.clone(), location: location.clone(), completion_time: d.clone() }
                    ))
                }
                Command::Claim(d) => {
                    Ok((
                        Self::Claimed { tracking_id: tracking_id.clone(), location: location.clone(), claimed_time: d.clone() },
                        Event::Claimed { tracking_id: tracking_id.clone(), completion_time: d.clone() }
                    ))
                }
                _ => invalid_state()
            }
            Self::OnBoardCarrier { tracking_id, voyage_no, .. } => match cmd {
                Command::Unload(l, d) => {
                    Ok((
                        Self::InPort { tracking_id: tracking_id.clone(), location: l.clone() },
                        Event::Unloaded { tracking_id: tracking_id.clone(), voyage_no: voyage_no.clone(), location: l.clone(), completion_time: d.clone() }
                    ))
                }
                _ => invalid_state()
            }
            Self::Claimed { .. } => invalid_state()
        }
    }
}

fn invalid_state<T>() -> DeliveryResult<T> {
    Err(CommandError::InvalidState)
}
